File: app/controllers/Wvyu.scala
package controllers

import play.api.mvc.{Action, Controller}
import data.ServiceClient
import play.api.libs.concurrent.Execution.Implicits._

/**
 * An example standalone module that represents the "Who's Viewed Your Updates" module from the LinkedIn homepage.
 */
class Wvyu(serviceClient: ServiceClient) extends Controller {

  def index(embed: Boolean) = Action.async { implicit request =>
    val likesCountFuture = serviceClient.makeServiceCall("likes")
    val commentsCountFuture = serviceClient.makeServiceCall("comments")

    for {
      likesCount <- likesCountFuture
      commentsCount <- commentsCountFuture
    } yield {
      if (embed) Ok(views.html.wvyu.wvyuBody(likesCount.toInt, commentsCount.toInt))
      else Ok(views.html.wvyu.wvyu(likesCount.toInt, commentsCount.toInt))
    }
  }
}

File: sample-app-java/app/controllers/Wvyu.java
package controllers;

import data.ServiceClient;
import play.libs.F;
import play.mvc.Controller;
import play.mvc.Result;

import javax.inject.Inject;

public class Wvyu extends Controller {
  private final ServiceClient serviceClient;

  @Inject
  public Wvyu(ServiceClient serviceClient) {
    this.serviceClient = serviceClient;
  }

  public F.Promise<Result> index(Boolean embed) {
    F.Promise<String> likesPromise = serviceClient.makeServiceCall("likes");
    F.Promise<String> commentsPromise = serviceClient.makeServiceCall("comments");

    return likesPromise.flatMap(likes -> commentsPromise.map(comments -> {
      int likesCount = Integer.parseInt(likes);
      int commentsCount = Integer.parseInt(comments);

      if (embed) {
        return ok(views.html.wvyu.wvyuBody.apply(likesCount, commentsCount));
      } else {
        return ok(views.html.wvyu.wvyu.apply(likesCount, commentsCount));
      }
    }));
  }
}